pkg/controllers/autopropagation/service_sync_controller.go

```diff
@@ -0,0 +1,187 @@
+func (c *ServiceController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	service := &corev1.Service{}
+	if err := c.Client.Get(ctx, req.NamespacedName, service); err != nil {
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	if !service.DeletionTimestamp.IsZero() {
+		// Do nothing, just return as we have added owner reference to PropagationPolicy.
+		// Work will be removed automatically by garbage collector.
+		return reconcile.Result{}, nil
+	}
+
+	annotations := service.GetAnnotations()
+	if annotations == nil {
+		klog.Warningf("service %q of namespace %q has no annotations. skip", service.Name, service.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	clusterList := &clusterv1alpha1.ClusterList{}
+	if err := c.Client.List(context.TODO(), clusterList); err != nil {
+		klog.Errorf("Failed to list clusters, error: %v", err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	deployClusters := c.syncClusters(service.Namespace, service.Name, annotations, clusterList.Items)
+	if deployClusters == nil {
+		return reconcile.Result{}, nil
+	}
+
+	if err := c.buildPropagationPolicy(service, deployClusters); err != nil {
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *ServiceController) syncClusters(namespace, service string, annotations map[string]string, clusters []clusterv1alpha1.Cluster) []string {
+	var deployClusters []string
+
+	if v, ok := annotations[annotationsGlobalKEYName]; ok && v == "true" {
+		klog.Infof("service %q propagation policy for namespace %q is global.", service, namespace)
+		for _, cluster := range clusters {
+			deployClusters = append(deployClusters, cluster.Name)
+		}
+		return deployClusters
+	}
+
+	v, ok := annotations[annotationsMembersKEYName]
+	if !ok {
+		klog.Infof("The propagation policy of service %q of namespace %q is not global, and no cluster is specified. Skip", service, namespace)
+		return nil
+	}
+
+	members := strings.Split(v, ",")
+	for _, cluster := range clusters {
+		for _, member := range members {
+			if cluster.Name == member {
+				deployClusters = append(deployClusters, cluster.Name)
+				continue
+			}
+		}
+	}
+
+	klog.V(4).Infof("member clusters: %v,valid number of deploy clusters: %v", members, deployClusters)
+	return deployClusters
+}
+
+// buildPropagationPolicy create service PropagationPolicy
+func (c *ServiceController) buildPropagationPolicy(service *corev1.Service, clusters []string) error {
+	pp := &policyv1alpha1.PropagationPolicy{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: policyv1alpha1.GroupVersion.String(),
+			Kind:       policyv1alpha1.ResourceKindPropagationPolicy,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      service.Name,
+			Namespace: service.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
```

Comment:
When the service object is deleted, the propagationpolicy  also deleted ？

Reply:
yes.